src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,574 @@
+        if (ls_res.has_failure()) {
+            co_return ls_res.as_failure();
+        }
+        subscan_res left_scan = ls_res.value();
+
+        // Right scan
+        auto rs_res = co_await subscan_right(segments, range.last, deadline);
+        if (rs_res.has_failure()) {
+            co_return rs_res.as_failure();
+        }
+        subscan_res right_scan = rs_res.value();
+        vassert(left_scan.done && right_scan.done, "Scan failed");
+
+        // compute size
+        size_t total_size = 0;
+        bool is_compacted = false;
+        if (segments.size() > 1) {
+            size_t mid_size = 0;
+            size_t ix_last = segments.size() - 1;
+            size_t left_size = segments.front()->size_bytes()
+                               - left_scan.filepos;
+            size_t right_size = right_scan.filepos;
+            for (size_t i = 0; i < segments.size(); i++) {
+                mid_size += segments[i]->size_bytes()
+                            * (i == 0 || i == ix_last ? 0 : 1);
+                is_compacted = is_compacted
+                               || segments[i]->finished_self_compaction();
+            }
+            total_size = left_size + mid_size + right_size;
+            vlog(
+              _ctxlog.debug,
+              "Computed size components: {}-{}-{}, total: {}, is_compacted: {}",
+              left_size,
+              mid_size,
+              right_size,
+              total_size,
+              is_compacted);
+        } else {
+            total_size = right_scan.filepos - left_scan.filepos;
+            is_compacted = is_compacted
+                           || segments.front()->finished_self_compaction();
+            vlog(
+              _ctxlog.debug,
+              "Computed size components: {}-{}, total: {}, is_compacted: {}",
+              left_scan.filepos,
+              right_scan.filepos,
+              total_size,
+              is_compacted);
+        }
+
+        //
+        cloud_upload_parameters result{
+          .size_bytes = total_size,
+          .is_compacted = is_compacted,
+          .offsets = range,
+          .first_real_offset = left_scan.base_offset,
+          .first_data_timestamp = left_scan.first_timestamp,
+          .last_real_offset = right_scan.last_offset,
+          .last_data_timestamp = right_scan.max_timestamp,
+          .segment_locks = std::move(holders),
+        };
+
+        co_return result;
+
+    } catch (...) {
+        if (ssx::is_shutdown_exception(std::current_exception())) {
+            vlog(
+              _ctxlog.debug,
+              "Shutdown exception: {}",
+              std::current_exception());
+            co_return make_error_code(error_outcome::shutting_down);
+        }
+        vlog(
+          _ctxlog.error, "Unexpected exception: {}", std::current_exception());
+        co_return make_error_code(error_outcome::unexpected_failure);
+    }
+}
+
+ss::future<result<segment_upload::subscan_res>> segment_upload::subscan(
```

Comment:
nit: gate and logger aside, feels like this should be a static storage utility method